query: new function for showing saved config

cathay4t opened the pull request:
Introduced `libnmstate.show_saved_config()` for querying
saved network configuration persistent after reboot.

The output will use the same schema of `libnmstate.show().

The `-s` and `--saved-config` option has been added to `show`
subcommand of nmstatectl.

The `io.nmstate.ShowSavedConfig` function has been added to varlink
interface of nmstatectl.

tyll commented on libnmstate/netinfo.py:
```diff
@@ -33,3 +35,13 @@ def show(*, include_status_data=False):
         return show_with_plugins(plugins, include_status_data)
+
+
+def show_running_config():
```
What is the difference between `show_running_config()` and `show()`?

cathay4t replied:
`show` is running state including auto state like DHCP/LLDP.

`show_running_config()` is configure only.

tyll commented on libnmstate/netinfo.py:
```diff
@@ -33,3 +35,13 @@ def show(*, include_status_data=False):
+        return show_running_config_with_plugins(plugins)
+
+
+def show_saved_config():
```
Usually, the configuration is called state in nmstate, so maybe use `state` here, too.

cathay4t replied:
`state` indicate full network status. `config` is for configuration only.

ffmancera replied:
I am fine with the name, but should it be an argument of show() instead of a new function?

cathay4t replied:
The `show()` is returning a state, while this function returning a config. So I prefer `show_saved_config()` over
`show(type=saved)` as the latter one require a constant.

cathay4t replied:
@EdDev What's your idea on this? Should we use new argument of `show()` or new function?

cathay4t replied:
It's been a while. And we already merged the `show_running_config()` PR #1419 .

cathay4t committed:
query: new function for showing saved config

Introduced `libnmstate.show_saved_config()` for querying
saved network configuration persistent after reboot.

The output will use the same schema of `libnmstate.show().

The `-s` and `--saved-config` option has been added to `show`
subcommand of nmstatectl.

The `io.nmstate.ShowSavedConfig` function has been added to varlink
interface of nmstatectl.

Integration test cases added by bring interface down a interface and the
`show_saved_config()` should include its configure as up state.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Due to NetworkManager bug: https://bugzilla.redhat.com/show_bug.cgi?id=1913551

when there is a memory only profile exists, nmstate cannot query out the on-disk profiles.

ffmancera commented on libnmstate/nm/profile.py:
```diff
@@ -575,3 +578,52 @@ def _is_memory_only(nm_profile):
+    NM.SETTING_OVS_BRIDGE_SETTING_NAME: InterfaceType.OVS_BRIDGE,
+    NM.SETTING_OVS_PORT_SETTING_NAME: InterfaceType.OVS_PORT,
+    NM.SETTING_OVS_INTERFACE_SETTING_NAME: InterfaceType.OVS_INTERFACE,
+    NM.SETTING_VETH_SETTING_NAME: InterfaceType.VETH,
```
In order to provide backward compatibility `NM.SETTING_VETH_SETTING_NAME` cannot be called directly. A private helper function it is needed as done in `libnmstate/nm/translator.py` and `libnmstate/nm/veth.py`

ffmancera commented on libnmstate/nm/veth.py:
```diff
@@ -89,3 +89,10 @@ def create_iface_for_nm_veth_peer(iface):
+
+
+def get_veth_config(nm_profile):
+    nm_setting = nm_profile.get_setting_by_name(NM.SETTING_VETH_SETTING_NAME)
```
Same here..

ffmancera reviewed (changes requested):
It looks good. There is just a nit about backward compatibility with veth support. Thanks!

cathay4t commented:
Closing as NetworkManager is hiding saved profile when there is a memory only profile exists:
https://bugzilla.redhat.com/show_bug.cgi?id=1913551

Will continue my work on this feature once NM fixed this problem.
